freegs/reconstruction.py

```
@@ -0,0 +1,803 @@
+        Equilibrium.__init__(self, **kwargs)
+        self.test = test
+
+        # Defining Reconstruction Parameters
+        self.pprime_order = pprime_order
+        self.ffprime_order = ffprime_order
+        self.use_VesselCurrents = use_VesselCurrents
+        self.use_VerticalControl = use_VerticalControl
+        self.use_CurrentInitialisation = use_CurrentInitialisation
+        self.tolerance = tolerance
+        self.coil_weight = coil_weight
+        self.show=show
+
+        self.dR = self.R[1, 0] - self.R[0, 0]
+        self.dZ = self.Z[0, 1] - self.Z[0, 0]
+
+        # Creating Reconstruction Grid and Greens functions
+        self.generate_Greens()
+
+        # Reconstruction attributes
+        self.psi_norm = None # normalised psi
+        self.c = None # coefficients
+        self.check_limited = True
+        self.pause = 0.0001
+
+    # Methods for building measurment and uncertainty matrices
+    def take_Measurements_from_dictionary(self, measurement_dict, sigma_dict):
+        """
+        Function called to generate measurement matrix and sigma
+        Also give new tokamak object the coil currents
+
+        Parameters
+        ----------
+        tokamak - tokamak object
+        measurement_dict - dictionary containing measurements
+        sigma_dict - dictionry containing uncertainty values
+
+        Returns
+        -------
+        M - Returns measurement matrix
+        sigma - Returns uncertainty values
+        """
+
+        M = []
+        sigma = []
+
+        # Giving sensor measurements to M
+        for sensor in self.tokamak.sensors:
+            M.append([measurement_dict[sensor.name]])
+            sigma.append([sigma_dict[sensor.name]])
+
+        # Giving coil currents to M
+        for name, coil in self.tokamak.coils:
+            M.append([measurement_dict[name]])
+            sigma.append([sigma_dict[name]])
+
```

doesnt ensure all measurements are used.  should raise an error in this case. 

also better to loop through measurements or loop through objects?

---

one of my tests is actually designed such that not all measurements are used. the reconstruction is based on the machine that is being used, so iff there are some coils in the measurement dictionary that don't actually exist on the machine that is being used for the reconstruction then including all measurements would mean that there would be no way the programme would be abe to reconstruct. of course, this does bank on the fact that the names of the coils and sensors in the dictionary are identical to that in the machine.